libsupport/include/katana/JSONTracer.h

```diff
@@ -0,0 +1,84 @@
+#ifndef KATANA_LIBSUPPORT_KATANA_JSONTRACER_H_
+#define KATANA_LIBSUPPORT_KATANA_JSONTRACER_H_
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "katana/ProgressTracer.h"
+#include "katana/config.h"
+
+namespace katana {
+
+class KATANA_EXPORT JSONTracer : public ProgressTracer {
+  JSONTracer(uint32_t host_id = 0, uint32_t num_hosts = 1)
+      : ProgressTracer(host_id, num_hosts) {}
+
+public:
+  static std::shared_ptr<JSONTracer> Make(
+      uint32_t host_id = 0, uint32_t num_hosts = 1);
+
+  std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, bool ignore_active_span) override;
+  std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name,
+      const std::shared_ptr<ProgressSpan>& child_of) override;
```

Comment:
Why do we need a class hierarchy over spans?

I can understand why we would want a variety of different output backends (JSON output, console output) but why would we need different implementations of a "progress span" or "progress context"? Any backend specific behavior can be predicated on the backend used not the span or context used.

Reply:
We'll wrap the OpenTracer Jaeger's span class with this for example, what do you mean by predicate?  Have a flag or callback?

Reply:
So "JSON{Tracer,Span,Context}" doesn't mean thing that produces JSON output but simply "tracing implementation that isn't OpenTracer Jaeger"? And the point is we want to support programs that have both OpenTracer Jaeger spans and our spans at the same time?

The point of a virtual class hierarchy is to allow extension and polymorphism within the same program.

If you just wanted abstraction, you can just as easily just use information hiding instead:

```cpp
class ProgressSpan {
  // Today OpenTracingSpan tomorrow WhizSpan
  OpenTracingSpan span_;

public: 
  int GetSpanID() const;
};
```

And the fact that there are a series of linked datatypes, `JSON{Tracer,Span,Context}` seems weird to me. What a span or logging context is should be independent of how it is rendered. If the idea is to support multiple implementations simultaneously, I can sorta see a reason for this but I still don't see why it couldn't just be `JSONTracer` `Span` and `Context`.

Reply:
My viewpoint was that we could use the JSONTracer for local development like through mpirun so we don't have to setup a Jaeger or other OpenTracing collector just to get logs while testing
Then for graph-workers we could have end-to-end tracing through a real OpenTracing implementation like Jaeger, maybe making it optional to get output to stdout through the JSONTracer stuff as well for just the graph-worker

I guess I don't really understand what you mean by a span or logging context should be independent of how it is rendered?  Maybe a better name would be MockTracer instead of JSONTracer?  It's a different tracing system than Jaeger regardless of how it is outputted?

Reply:
My point, boiled down, is that spans and contexts are part of the API of tracing. They say things like "this is a region of interest", "this region is contained in another region". Whether we have a mock backend or a real one, the regions are the same and their interpretation is the same. So, I don't (yet) understand why we would have different types of spans related to what backend we use.

To put it another way, why couldn't the mock backend and the open tracing backend take in the same definition of a span?

Reply:
Those are abstract interfaces in OpenTracing's API as well
https://github.com/opentracing/opentracing-cpp/blob/master/include/opentracing/span.h

Reply:
I guess another way of looking at it is:
As far as I know, no OpenTracing implementation actually builds a complete span and then sends it
What they do is send intermediary span data (like when you start/finish/log/add a tag) to a collector
Then the collector assembles the intermediary span data into complete spans and traces

My implementation does the same, that's why I added a python script to assemble the tracer output into a Jaeger compatible format

Reply:
So spans are an abstract concept and their implementations vary widely, so they need a common interface

Reply:
For contexts, those follow the same basic reasoning since OpenTracing allows "baggage items" that are propagated to every child span once they are added.  Then there is a description language for how to handle adding to baggage items if say "sum" is already present on it, then setting "sum" to 42 would add 42 to the current sum, or if it is a string maybe you want to concatenate
Then there is also the concept of dropping baggage items if they become too large since these need to be propagated to every child span so there is logic there as well

Reply:
https://github.com/jaegertracing/jaeger-client-cpp/blob/master/src/jaegertracing/Span.h

Reply:
Actually I was wrong about saying all tracer implementations output intermediary span details
Looks like Jaeger at least waits until it has a complete span to report it

Reply:
Ok.

> As far as I know, no OpenTracing implementation actually builds a complete span and then sends it
> What they do is send intermediary span data (like when you start/finish/log/add a tag) to a collector
> Then the collector assembles the intermediary span data into complete spans and traces

Is a collector here something like JSONTracer or something else? In any event, these different ways of aggregating data seem seem orthogonal to the fact that the output will be JSON, which is the source of my confusion.

It also seems to lead to hacks like friend, which do not play well with class hierarchies.

If the intent is mirror the OpenTracing API, this virtual span business seems fine, but there are several differences with the OpenTracing API. For example, the OpenTracing API doesn't have any of this shared_ptr business for `StartSpan`: https://github.com/opentracing/opentracing-cpp/blob/master/include/opentracing/tracer.h#L92

Can you summarize why we need these differences?

Reply:
Mostly because I added ProgressScope for RAII and allow getting the active span
To be honest I've been looking mostly at the python API at:
https://opentracing-python.readthedocs.io/en/latest/api.html
So I am a bit unfamiliar with the C++ API, looking at Jaeger's examples and from what I can tell, C++'s OpenTRacing API does not support getting the active span so any time you want to create a child span or add span data you need to pass along the span like in
https://github.com/jaegertracing/jaeger-client-cpp/blob/master/examples/App.cpp
In my API we do not need to pass anything to get the same hierarchy, like in the python API

A collector would be run on a separate host in a cluster and aggregate span data-storing it in a database maybe or just making it available to be viewed at a certain web address

Reply:
Would you prefer I just close this and use the OpenTracing API instead?

Reply:
The C++ API lacks a main part of the OpenTracing API that I am trying to provide by adding a concept of an activespan and scopes
https://opentracing.io/docs/overview/scopes-and-threading/